File: app/auth/modules/custom_headers.py
from __future__ import annotations

import os
from typing import Any

import httpx
from redis.asyncio import Redis

from app.auth.base import AuthContext, AuthResult, _invalidate_on_from_config


class CustomHeadersAuth:
    """Inject arbitrary static headers sourced from env vars or literal values."""

    name = "custom_headers"

    def __init__(self, headers: dict[str, str], invalidate_on: set[int]) -> None:
        self._headers = headers
        self._invalidate_on = invalidate_on

    @classmethod
    def from_config(
        cls, config: dict[str, Any], *, redis: Redis, http: httpx.AsyncClient
    ) -> CustomHeadersAuth:
        raw_headers: dict[str, Any] = config.get("headers", {})
        resolved: dict[str, str] = {}
        for header_name, spec in raw_headers.items():
            if "value" in spec:
                resolved[header_name] = spec["value"]
            elif "env" in spec:
                env_name = spec["env"]
                val = os.environ.get(env_name)
                if not val:
                    raise ValueError(
                        f"CustomHeadersAuth: env var {env_name!r} for header {header_name!r} "
                        "is not set or empty."
                    )
                resolved[header_name] = val
            else:
                raise ValueError(
                    f"CustomHeadersAuth: header {header_name!r} must have 'env' or 'value'."
                )
        return cls(
            headers=resolved,
            invalidate_on=_invalidate_on_from_config(config),
        )

    async def apply(self, ctx: AuthContext) -> AuthResult:
        return AuthResult(headers=dict(self._headers))

    async def invalidate(self) -> None:
        pass

    def is_rejection(self, status_code: int, body: dict | None) -> bool:
        return status_code in self._invalidate_on

File: app/ts/src/auth/modules/custom_headers.ts
import type { AuthContext, AuthResult } from "../../core/types.ts";
import type { AuthModule, AuthDeps } from "../types.ts";
import { invalidateOnFromConfig } from "../types.ts";

export class CustomHeadersAuth implements AuthModule {
  static readonly typeName = "custom_headers";

  // Headers resolved once at construction time.
  private readonly headers: Record<string, string>;
  private readonly invalidateOn: Set<number>;

  private constructor(
    headers: Record<string, string>,
    invalidateOn: Set<number>,
  ) {
    this.headers = headers;
    this.invalidateOn = invalidateOn;
  }

  static fromConfig(
    config: Record<string, unknown>,
    deps: AuthDeps,
  ): CustomHeadersAuth {
    const rawHeaders = config["headers"];
    if (typeof rawHeaders !== "object" || rawHeaders === null) {
      throw new Error(
        "CustomHeadersAuth: 'headers' config field is required and must be an object.",
      );
    }

    const resolved: Record<string, string> = {};

    for (const [headerName, spec] of Object.entries(
      rawHeaders as Record<string, unknown>,
    )) {
      if (typeof spec !== "object" || spec === null) {
        throw new Error(
          `CustomHeadersAuth: header '${headerName}' spec must be an object.`,
        );
      }
      const specObj = spec as Record<string, unknown>;

      if (typeof specObj["value"] === "string") {
        resolved[headerName] = specObj["value"];
      } else if (typeof specObj["env"] === "string") {
        const envName = specObj["env"];
        const val = deps.env[envName];
        if (typeof val !== "string" || !val) {
          throw new Error(
            `CustomHeadersAuth: env var '${envName}' for header '${headerName}' is not set or empty.`,
          );
        }
        resolved[headerName] = val;
      } else {
        throw new Error(
          `CustomHeadersAuth: header '${headerName}' must have 'env' or 'value'.`,
        );
      }
    }

    // Default invalidate_on is empty — custom header injection rarely has a clear rejection signal.
    const invalidateOn = invalidateOnFromConfig(config, []);

    return new CustomHeadersAuth(resolved, invalidateOn);
  }

  async apply(_ctx: AuthContext): Promise<AuthResult> {
    return { headers: { ...this.headers } };
  }

  async invalidate(): Promise<void> {}

  isRejection(statusCode: number, _body: unknown): boolean {
    return this.invalidateOn.has(statusCode);
  }
}
